# 4.6.1. Merging Data - Example

## This script contain following points:

### 1. Import the data set into Jupyter as a new dataframe: df_ords_prior

### 2. Check the dimensions of the imported dataframe

### 3. Merge the dataframe with df_ords dataframe

### 4. Use the indicator argument to check whether there was a full match between the two dataframes

## 01. Importing libraries

In [19]:
# Import libraries
import pandas as pd
import numpy as np
import os

## 02. Import the data sets into Jupyter

In [20]:
# Create path
path = r'C:\Users\Quinn\OneDrive\Documents\CF - Data Analyst\Data Immersion\Instacart Basket Analysis'

In [21]:
# Import orders_products_prior dataset
df_ords_prior = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'orders_products_prior.csv'), index_col = False)

In [22]:
# Import checked orders dataset
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_checked.csv'), index_col = 0)

## 02. Check the dimension of the imported dataframe

In [23]:
# Check the output of df_ords_prior
df_ords_prior.head()

order_id  product_id  add_to_cart_order  reordered
0         2       33120                  1          1
1         2       28985                  2          1
2         2        9327                  3          0
3         2       45918                  4          1
4         2       30035                  5          0

In [24]:
# Check the output of df_ords
df_ords.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_last_order      first_order  
0                    NaN      First Order  
1                   15.0  Repeat Customer  
2                   21.0  Repeat Customer  
3                   29.0  Repeat Customer  
4                   28.0  Repeat Customer

In [25]:
df_ords_prior.shape

(32434489, 4)

In [26]:
df_ords.shape

(3421083, 7)

## 03. Merge the dataframe with df_ords dataframe

In [27]:
# Merge both dataframe using the mutual column order_id
df_merged_large = df_ords.merge(df_ords_prior, on = 'order_id', indicator = True)

In [28]:
df_merged_large.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_last_order  first_order  product_id  add_to_cart_order  \
0                    NaN  First Order         196                  1   
1                    NaN  First Order       14084                  2   
2                    NaN  First Order       12427                  3   
3                    NaN  First Order       26088                  4   
4                    NaN  First Order       26405                  5   

   reordered _merge  
0          0   both  
1          0   both  
2          0   both  
3          0   both  
4          0   both

## 04. Use the indicator argument to check whether there was a full match between the two dataframes

In [29]:
df_merged_large['_merge'].value_counts()

_merge
both          32434489
left_only            0
right_only           0
Name: count, dtype: int64

### The resulting table will only contain observations found in both dataframes, which are 32,434,489. We used inner join for the purpose of this project.

## 05. Exporting data

In [30]:
# Export data to csv
df_merged_large.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_products_combined.csv'))

In [31]:
# Export data to pkl
df_merged_large.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_combined.pkl'))